ouroboros-consensus-protocol/src/Ouroboros/Consensus/Protocol/Praos/VRF.hs

```diff
@@ -0,0 +1,126 @@
+{-# LANGUAGE DataKinds                  #-}
+{-# LANGUAGE DeriveGeneric              #-}
+{-# LANGUAGE DerivingStrategies         #-}
+{-# LANGUAGE GADTs                      #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE KindSignatures             #-}
+{-# LANGUAGE OverloadedStrings          #-}
+{-# LANGUAGE RankNTypes                 #-}
+{-# LANGUAGE ScopedTypeVariables        #-}
+{-# LANGUAGE TypeApplications           #-}
+{-# OPTIONS_GHC -Wno-unticked-promoted-constructors #-}
+
+module Ouroboros.Consensus.Protocol.Praos.VRF (
+    UnifiedVRF
+  , VRFUsage (..)
+  , hashVRF
+  , mkUnifiedVRF
+  , vrfLeaderCheckValue
+  , vrfNonceValue
+  ) where
+
+import           Cardano.Binary (ToCBOR)
+import           Cardano.Crypto.Hash (Blake2b_256, Hash, castHash, hashToBytes,
+                     hashWith)
+import qualified Cardano.Crypto.Hash as Hash
+import           Cardano.Crypto.Util
+                     (SignableRepresentation (getSignableRepresentation))
+import           Cardano.Crypto.VRF (OutputVRF (..), getOutputVRFBytes)
+import           Cardano.Ledger.BaseTypes (Nonce (NeutralNonce, Nonce))
+import           Cardano.Ledger.Crypto (Crypto (HASH, VRF))
+import           Cardano.Ledger.Serialization (runByteBuilder)
+import           Cardano.Ledger.Slot (SlotNo (SlotNo))
+import qualified Data.ByteString.Builder as BS
+import qualified Data.ByteString.Builder.Extra as BS
+import           GHC.Generics (Generic)
+import           NoThunks.Class (NoThunks)
+
+-- | Newtype wrapper for a unified VRF value which may then be converted to
+-- either of:
+--
+-- - Leader selection value
+-- - Updated Nonce
+newtype UnifiedVRF = UnifiedVRF {unUnifiedVRF :: Hash Blake2b_256 UnifiedVRF}
```

Comment:
So I want to suggest some name changes and API changes here, that will make it clearer what is going on, and avoid bugs (such as ones we already made).

In the unified / range-extended VRF scheme, I think we need to distinguish (using types) between three things:
1. the VRF input (made from the slot & epoch nonce)
2. the raw VRF output (as put into the block header)
3. the cooked VRF outputs, used for leader election and nonce update

At the moment the API doesn't properly distinguish 2 and 3. Currently both have type `OutputVRF`. The function to get 3 from 2 has type `OutputVRF -> OutputVRF`, and this makes it easy to forget how many times this function has been applied. The current bug is that it's applied twice, but one could easily make the reverse mistake. If we properly distinguish the 3 types then we cannot go wrong easily.

Also, I want to grumble about the name `UnifiedVRF`. This is really the VRF input. So lets call it that, or something like that. From a first reading one cannot guess if it's the input or output (I'd have guessed output).

I'm not sure we need the term "unified" in the API functions. Though perhaps in the module name.

Can we avoid exporting `hashVRF`, and just export `vrfLeaderCheckValue` and `vrfNonceValue`? Then once those two have different input and output types, then it becomes pretty hard to get anything wrong.

So we'll want names for the "raw" and "cooked" VRF outputs. That's not a concrete suggestion, I think I'm just hungry. :plate_with_cutlery:

Reply:
I originally went down this route, and abandoned it because the ledger expects the "raw" outputVRF. But this is a very strong argument that it's worth the hassle of changing the ledger, so I've done just that.